php-checks/src/main/java/org/sonar/php/checks/UnusedLocalVariableCheck.java

```diff
@@ -19,144 +19,117 @@
-    AstNode parent = varIdentifier.getParent();
-    return parent.getParent().is(PHPGrammar.GLOBAL_VAR) || parent.is(PHPGrammar.STATIC_VAR, PHPGrammar.LEXICAL_VAR);
+    super.visitFunctionExpression(tree);
   }
 
   @Override
-  public void leaveNode(AstNode astNode) {
-    if (astNode.is(FunctionUtils.functions())) {
-      reportUnusedVariable();
-      scopes.pop();
+  public void visitFunctionCall(FunctionCallTree tree) {
+    for (ExpressionTree argument : tree.arguments()) {
+      if (argument.is(Tree.Kind.VARIABLE_IDENTIFIER)) {
+        exclusions.add((IdentifierTree) argument);
+      }
     }
+    super.visitFunctionCall(tree);
   }
 
-  private void declareNewLocalVariable(AstNode astNode) {
-    AstNode leftExpr = getLeftHandExpression(astNode);
-    if (leftExpr != null && leftExpr.is(PHPGrammar.VARIABLE_WITHOUT_OBJECTS)) {
-      getCurrentScope().declareVariable(leftExpr);
-    }
+  @Override
+  public void visitCatchBlock(CatchBlockTree tree) {
+    exclusions.add(tree.variable());
+    super.visitCatchBlock(tree);
   }
 
-  private AstNode getLeftHandExpression(AstNode assignmentExpr) {
-    AstNode leftExpr = assignmentExpr.getFirstChild();
+  @Override
+  public void visitForEachStatement(ForEachStatementTree tree) {
+    if (tree.key() != null) {
+      if (tree.value().is(Tree.Kind.VARIABLE_IDENTIFIER)) {
+        exclusions.add((IdentifierTree) tree.value());
 
-    if (leftExpr.is(PHPGrammar.MEMBER_EXPRESSION)) {
-      return leftExpr.getFirstChild();
-    } else if (leftExpr.is(PHPGrammar.POSTFIX_EXPR)) {
-      return leftExpr.getFirstChild().getFirstChild();
-    } else {
-      return null;
+      } else if (tree.value().is(Tree.Kind.REFERENCE_VARIABLE) && ((ReferenceVariableTree) tree.value()).variableExpression().is(Tree.Kind.VARIABLE_IDENTIFIER)) {
+        exclusions.add((IdentifierTree) ((ReferenceVariableTree) tree.value()).variableExpression());
+      }
     }
+    super.visitForEachStatement(tree);
   }
 
-  private void reportUnusedVariable() {
-    for (Variable localVar : getCurrentScope().getLocalVariables().values()) {
-
-      if (localVar.getUsage() == 1) {
-        getContext().createLineViolation(this, "Remove this unused \"{0}\" local variable.", localVar.getDeclaration(),
-          localVar.getDeclaration().getTokenOriginalValue());
+  @Override
+  public void visitCompilationUnit(CompilationUnitTree tree) {
+    exclusions.clear();
+    super.visitCompilationUnit(tree);
+    for (Scope scope : context().symbolTable().getScopes()) {
+      if (scope.tree() instanceof FunctionTree) {
+        checkScope(scope);
       }
     }
   }
 
-  private LocalVariableScope getCurrentScope() {
-    return scopes.peek();
+  private void checkScope(Scope scope) {
+    for (Symbol symbol : scope.getSymbols(Kind.VARIABLE)) {
+      // symbol should be declared in this scope
+      if (symbol.scope().equals(scope)) {
+        if (symbol.usages().isEmpty() && !exclusions.contains(symbol.declaration())) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1066)